spec/unit/provider/package/dpkg_spec.rb

```diff
@@ -68,105 +71,232 @@ def dpkg_query_execution_returns(output)
+          dpkg_query_execution_with_multiple_args_returns(dpkg_query_result.gsub("installed","config-files"), args, query_args)
+          expect(provider.query[:ensure]).to eq(:absent)
+        end
+
+        it "considers the package absent if it is marked 'half-installed'" do      
+          dpkg_query_execution_with_multiple_args_returns(query_output.gsub("installed","half-installed"),args_with_provides,virtual_packages_query_args)
+          dpkg_query_execution_with_multiple_args_returns(dpkg_query_result.gsub("installed","half-installed"), args, query_args)
+          expect(provider.query[:ensure]).to eq(:absent)
+        end
+        it "considers the package absent if it is marked 'unpacked'" do       
+          dpkg_query_execution_with_multiple_args_returns(query_output.gsub("installed","unpacked"),args_with_provides,virtual_packages_query_args)
+          dpkg_query_execution_with_multiple_args_returns(dpkg_query_result.gsub("installed","unpacked"), args, query_args)
+          expect(provider.query[:ensure]).to eq(:absent)
+        end
+
+        it "considers the package absent if it is marked 'half-configured'" do        
+          dpkg_query_execution_with_multiple_args_returns(query_output.gsub("installed","half-configured"),args_with_provides,virtual_packages_query_args) 
+          dpkg_query_execution_with_multiple_args_returns(dpkg_query_result.gsub("installed","half-configured"), args, query_args)     
+          expect(provider.query[:ensure]).to eq(:absent)
+        end
+
+        it "considers the package held if its state is 'hold'" do        
+          dpkg_query_execution_with_multiple_args_returns(query_output.gsub("install","hold"),args_with_provides,virtual_packages_query_args)
+          dpkg_query_execution_with_multiple_args_returns(dpkg_query_result.gsub("install","hold"), args, query_args)
+          expect(provider.query[:ensure]).to eq(:held)
+        end
+        context "regex check for query search" do
+          let(:resource_name) { 'python-email' }
+          let(:resource) { instance_double('Puppet::Type::Package') }
+          before do
+            allow(resource).to receive(:[]).with(:name).and_return(resource_name)
+            allow(resource).to receive(:[]=)
+          end
+
+          it "checks if virtual package regex for query is correctand phisical package is installed" do
+            dpkg_query_execution_with_multiple_args_returns(query_output,args_with_provides,virtual_packages_query_args)
+            dpkg_query_execution_with_multiple_args_returns(dpkg_query_result, args, query_args)
+            expect(provider.query).to match({:desired=>"install", :ensure=>"2.7.13", :error=>"ok", :name=>"python", :provider=>:dpkg, :status=>"installed"})
+          end
+          
+          context "regex check with no partial matching" do
+            let(:resource_name) { 'python-em' }
```

Comment:
Given that python-em works when escaped or not:

```
irb(main):016:0> "python-em" =~ /#{Regexp.escape("python-em")}/
=> 0
irb(main):017:0> "python-em" =~ /python-em/
=> 0
```

I think it's better to use (or add a different test) for `g++` since we had that regression:

```
irb(main):025:0> "g".match(/g++/)
=> #<MatchData "g">
irb(main):026:0> "g".match(/g\+\+/)
=> nil
irb(main):027:0> "g".match(/#{Regexp.escape("g++")}/)
=> nil
irb(main):028:0> "g++".match(/#{Regexp.escape("g++")}/)
=> #<MatchData "g++">
```

Reply:
Added test, see in:
https://github.com/puppetlabs/puppet/pull/7701/files#diff-68fbdeeded08eb112e8ce141c005d7beR96